agent/health_endpoint.go

```diff
@@ -220,35 +219,19 @@ func (s *HTTPHandlers) healthServiceNodes(resp http.ResponseWriter, req *http.Re
 		return nil, nil
 	}
 
-	// Make the RPC request
-	var out structs.IndexedCheckServiceNodes
-	defer setMeta(resp, &out.QueryMeta)
+	// TODO: handle this for all endpoints in parseConsistency
+	args.QueryOptions.UseCache = s.agent.config.HTTPUseCache && args.QueryOptions.UseCache
 
-	if s.agent.config.HTTPUseCache && args.QueryOptions.UseCache {
-		raw, m, err := s.agent.cache.Get(req.Context(), cachetype.HealthServicesName, &args)
-		if err != nil {
-			return nil, err
-		}
-		defer setCacheMeta(resp, &m)
-		reply, ok := raw.(*structs.IndexedCheckServiceNodes)
-		if !ok {
-			// This should never happen, but we want to protect against panics
-			return nil, fmt.Errorf("internal error: response type not correct")
-		}
-		out = *reply
-	} else {
-	RETRY_ONCE:
-		if err := s.agent.RPC("Health.ServiceNodes", &args, &out); err != nil {
-			return nil, err
-		}
-		if args.QueryOptions.AllowStale && args.MaxStaleDuration > 0 && args.MaxStaleDuration < out.LastContact {
-			args.AllowStale = false
-			args.MaxStaleDuration = 0
-			goto RETRY_ONCE
-		}
+	out, md, err := s.agent.rpcClientHealth.ServiceNodes(req.Context(), args)
+	if err != nil {
+		return nil, err
 	}
+
+	setCacheMeta(resp, &md)
```

Comment:
Ah, I think there might be a slight change in behaviour here. Previously the cache headers would not be set. I think I need to make another change so that this is only called when `UseCache` is set.